tap_harvest/client.py

```diff
@@ -0,0 +1,97 @@
+
+import backoff
+import requests
+import pendulum
+from singer import utils
+import singer
+
+LOGGER = singer.get_logger()
+
+BASE_ID_URL = "https://id.getharvest.com/api/v2/"
+BASE_API_URL = "https://api.harvestapp.com/v2/"
+
+class HarvestClient:#pylint: disable=too-many-instance-attributes
+    def __init__(self, client_id, client_secret, refresh_token, user_agent):
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._refresh_token = refresh_token
+        self._user_agent = user_agent
+        self._account_id = None
+        self.session = requests.Session()
+        self._refresh_access_token()
+
+    @backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        max_tries=5,
+        giveup=lambda e: e.response is not None and 400 <= e.response.status_code < 500,
+        factor=2)
+    def _make_refresh_token_request(self):
+        return requests.request('POST',
+                                url=BASE_ID_URL + 'oauth2/token',
+                                data={
+                                    'client_id': self._client_id,
+                                    'client_secret': self._client_secret,
+                                    'refresh_token': self._refresh_token,
+                                    'grant_type': 'refresh_token',
+                                },
+                                headers={"User-Agent": self._user_agent})
+
+    def _refresh_access_token(self):
+        LOGGER.info("Refreshing access token")
+        resp = self._make_refresh_token_request()
+        expires_in_seconds = resp.json().get('expires_in', 17 * 60 * 60)
+        self._expires_at = pendulum.now().add(seconds=expires_in_seconds)
+        resp_json = {}
+        try:
+            resp_json = resp.json()
+            self._access_token = resp_json['access_token']
+        except KeyError as key_err:
```

Comment:
@savan-chovatiya Please check the scenario of status code=500. Is there a possibility of getting exception other than 'KeyError'?

Reply:
We are raising exceptions before this in the `_make_refresh_token_request()' function for status codes other than 200 and 201 so no possibility of getting exceptions other than KeyError for the success status code.